zkevm-circuits/src/blake2b/circuit/src/lib.rs

```diff
@@ -0,0 +1,1855 @@
+    // by the current GeneralCell instance and the specified assigment row
+    pub fn assign(&self, region: &mut Region<F>, row: usize, value: Value<F>) -> Result<(), Error> {
+        region.assign_advice(|| "", self.column, shift_row(row, self.offset), || value)?;
+        Ok(())
+    }
+}
+
+// Describes the relative place of a combined selector
+#[derive(Copy, Clone)]
+pub struct Combiselector<F:FieldExt> {
+    allowed: Column<Fixed>, // The "allowed" column containing the fixed cell of the combined selector
+    selector: Column<Advice>, // The column containing the advice cell of the combined selector
+    offset: i32, // The combined selector's offset relative to the assignment row
+    _marker: PhantomData<F> // The marker used to specify the native field of the columns
+}
+
+impl<F:FieldExt> Combiselector<F> {
+    // Creates a Combiselector for the specified "allowed" column, 
+    // advice column and offset relative to the assignment row
+    pub fn new(allowed: Column<Fixed>, selector: Column<Advice>, offset: i32) -> Self {
+        Self { allowed, selector, offset, _marker: PhantomData }
+    }
+
+    // Creates the instance of Expression for the Combiselector
+    pub fn expr(&self, meta: &mut VirtualCells<F>) -> Expression<F> {
+        meta.query_fixed(self.allowed, Rotation(self.offset)) * meta.query_advice(self.selector, Rotation(self.offset))
+    }
+
+    // Activates or inactivates (depending on "active") the combined selector, which is    
+    // described by the current Combiselector instance and the specified assigment row
+    pub fn enable(&self, region: &mut Region<F>, row: usize, active: bool) -> Result<(), Error> {
+        let offset = shift_row(row, self.offset);
+        region.assign_advice(|| "", self.selector, offset, || known::<F>(active as u64))?;
+        Ok(())
+    }
+}
+
+type OctaChunk<F> = Chunk<F,8>;
+type ByteChunk<F> = Chunk<F,1>; 
+type BitCell<F> = ShortCell<F,1>;
+type SeptaCell<F> = ShortCell<F,127>;
+
+// Defines the gate class, whose representatives activate  
+// or inactivate the specified target combined selector
+#[derive(Clone)]
+pub struct SelectorGate<F:FieldExt> {
+    selectors: Vec<Combiselector<F>>, // The control Combiselectors 
+    input: bool, // The flag specifying whether the target combined selector should be activated
+    result: Combiselector<F> // The target Combiselector
+}
+
+impl<F:FieldExt> SelectorGate<F> {
+    fn configure(meta: &mut ConstraintSystem<F>,
+                 // The control Combiselectors 
+                 selectors: &[Combiselector<F>],
+                 // The flag specifying whether the target combined selector should be activated 
+                 input: bool,
+                 // The target Combiselector 
+                 result: Combiselector<F>) -> Self {
+        meta.create_gate("SelectorGate", |meta| {
+            vec![combine_selectors(meta, selectors) * (result.expr(meta) - gf(input as u128))]
+        });
```

Comment:
Not related to this PR, but there is something odd with `rustfmt` as this indentation is not a canonical format

Reply:
update: seems trailing spaces are breaking rustfmt, that's why it isn't working for this snippet. Consider removing trailing spaces

Reply:
Thanks, I will remove them.